composer/models/base.py

```diff
@@ -49,15 +49,30 @@ def loss(self, outputs, batch):
                 # pass batches and `forward` outputs to the loss
                 _, targets = batch
                 return F.cross_entropy(outputs, targets)
-
-    Attributes:
-        logger (Optional[Logger]): The :class:`.Logger`. The trainer sets the :class:`.Logger` on the
-            :attr:`~composer.core.event.Event.INIT` event.
     """
 
     def __init__(self) -> None:
         super().__init__()
-        self.logger: Optional[Logger] = None
+
+    def _get_logger(self) -> Optional[Logger]:
+        # The trainer monkey-patches this getter to set the logger.
+        # Since properties are bound to the class (and not the instance), the `logger` @property cannot be monkey-patched
+        # directly. Hence, this helper getter, which the @property(logger) uses below, is monkey-patched.
+        # This trick preserves the ability to `deepcopy` the model (which is required by some algorithms, such as SWA),
+        # as the Logger cannot be deepcopied. (The Logger is attached to the state, and the state has file objects).
+        # This trick works because methods are not copied but are instead re-bound by reference.
+        return None
+
+    @property
+    def logger(self) -> Optional[Logger]:
```

Comment:
well retrieving the logger still work after SWA deep copies the model? (which happens after the patch happens)

Reply:
Also, would it be better for the model to implement a `__deepcopy__()` that skips the logger attribute?

e.g. see https://docs.python.org/3/library/copy.html

Reply:
> well retrieving the logger still work after SWA deep copies the model? (which happens after the patch happens)

Yes, added a test case for this.